### PR TITLE
[MGPG-81] check format in constructor

### PR DESCRIPTION
**Pull request by Syquel**
Provides an refactored implementation of GpgVersion to address [MGPG-81](https://issues.apache.org/jira/projects/MGPG/issues/MGPG-81).

The parsing logic is being moved from "GpgVersion#compareTo" to "GpgVersion#parse" to throw IllegalArgumentException in a semantically correct way.

### CODE REVIEW

**Review by elharo — COMMENTED** [review]
can you split this into individual PRs for each issue? That makes it much easier to review and track in Jira and release notes.

File: src/test/java/org/apache/maven/plugins/gpg/GpgVersionTest.java
Comment (elharo) [inline src/test/java/org/apache/maven/plugins/gpg/GpgVersionTest.java:55]: hash code should be tested too

File: src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:75]: can this be a simple int comparison? e.g.

```
if (thisSegments[index] != otherSegments[index]) {
  return thisSegments[index] - otherSegments[index]
}
```

(Double check my order there)

**Comment by Syquel** [comment]
> 
> 
> can you split this into individual PRs for each issue? That makes it much easier to review and track in Jira and release notes.

I will reduce this PR to MGPG-81 by removing GpgVersion#equals and GpgVersion#hashCode first and submit another PR for MGPG-80 later on after this one is merged.

File: src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java
Comment (Syquel) [reply src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:75]: I think it would be preferable to adhere to the contract of Integer#compareTo / Integer#compare, which is internally implemented as `(x < y) ? -1 : ((x == y) ? 0 : 1)`, because this class is in my opinion just a container for an int array and should behave as such.

But your solution works of course, too. Your call.

Comment (elharo) [reply src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:74]: Yes. We should adhere to the contract. The contract is more lenient than the current implementation. It just requires negative, positive, or zero at the appropriate places, not -1, 0, or +1 specifically.

File: src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:45]: Where does this version string come from? That is, what is it the version of? Is it ever anything that does not match this pattern? E.g. 3.45.2-beta?

File: src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:47]: The regex might be overkill. Looks like you can just catch and convert the number format exception below.

File: src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java
Comment (Syquel) [reply src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:45]: As far as I can see the original author of this class made the assumption that it will always be a sequence of numbers delimited by dots (although the original author wrote in the class-level comment, that this is a SemVer parser).
The raw version string is read from the output of `gpg --version` apparently.

Unfortunately I am in no position to evaluate whether GPG will continue using this versioning scheme, but it seems they have consistently used it up until now: https://files.gpg4win.org/
In my opinion we can make the assumption, that GPG will continue using a three segment wide versioning scheme just as the original author implemented it.

File: src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java
Comment (Syquel) [reply src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:47]: The original author implemented it in a way, that the full output of `gpg --version` (e.g. `gpg (GnuPG) 2.2.1` or `gpg (GnuPG) 2.0.26 (Gpg4win 2.2.3)`) is passed to this function.
They used the regex to extract the version part out of that output.

It is debatable whether that is the best way to extract the version number (personally I would have extracted the version number part already in [GpgVersionParser.GpgVersionConsumer](https://github.com/apache/maven-gpg-plugin/blob/master/src/main/java/org/apache/maven/plugins/gpg/GpgVersionParser.java#L113) before passing it in here), but I would like to avoid introducing new bugs by breaking with the existing external contract of this class.

In the end it's your call again and I could widen the scope of the refactoring to the whole GPG version parser.

**Review by elharo — APPROVED** [review]

**Comment by elharo** [comment]
give me a minute to run this through jenkins

**Comment by elharo** [comment]
https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-gpg-plugin/job/check/

**Comment by Syquel** [comment]
Thanks for merging.

As promised I submitted PR #12 for MGPG-80.

File: src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java
Comment (Tibor17) [inline src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:36]: @Syquel 
@elharo 
How many times these patterns exist in the project?

Comment (Syquel) [reply src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:36]: There are two patterns AFAIK: One in [GpgVersionParser.GpgVersionConsumer](https://github.com/apache/maven-gpg-plugin/blob/master/src/main/java/org/apache/maven/plugins/gpg/GpgVersionParser.java#L113) and this one here.
This one could possibly be removed and `GpgVersionParser.GpgVersionConsumer` could pass the extracted version number to `GpgVersion`.

Comment (Syquel) [reply src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:36]: It might be even better to use `gpgconf --query-swdb gnupg` instead of `gpg --version`, because gpgconf returns [machine-readable output](https://github.com/gpg/gnupg/blob/master/tools/gpgconf.c#L360).

> $ gpgconf --query-swdb gnupg
> gnupg:2.2.27:-::32849:::::::

although I don't know if that would work with all distributions.

Comment (bmarwell) [reply src/main/java/org/apache/maven/plugins/gpg/GpgVersion.java:36]: Works on mac, too: 
```
$ gpgconf --query-swdb gnupg
gnupg:2.2.21:-::32849:::::::
```

**Comment by jira-importer** [comment]
Resolve #219

**Comment by jira-importer** [comment]
Resolve #223

**Comment by jira-importer** [comment]
Resolve #219

**Pull request merged**
